Add warning for dict.keys(),values(), items().sort

Eddy114514 opened the pull request:
Add a warning for calls to dict.keys().sort(), dict.values().sort(), and dict.items().sort(), informing users that in Python 3 these methods return iterable views rather than lists.
 
Add warning for call of .reverse() for keys(), values(), items(). Add attribute reverse() and point to `__reversed__()`.

Since directly calling .sort() on a dict view object in Python 3 would otherwise raise an AttributeError, the implementations of _PyObject_GetMethod and PyObject_GenericGetAttr were modified to detect this pattern. Instead of failing, they now return a dummy function that issues a warning and performs no action.

In Python 2, dict.keys(), dict.values(), and dict.items() returned lists, so calling .sort() directly was valid and modified the returned list in place. Replicating this behavior in Python 3 would require converting between view objects and lists, which add complexity.

Therefore, I decide to just warn the user about the semantic change without attempting to preserve the original sorting behavior.

<img width="1767" height="499" alt="image" src="https://github.com/user-attachments/assets/16a04e20-33d9-4d56-be9f-99219e8fb75b" />

ltratt commented on Objects/object.c:
```diff
@@ -1197,6 +1209,18 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
+        };
+        *method = PyCFunction_NewEx(&noop_def, obj, NULL);
+        return 0;
+        }
```
I think the formatting of this line is off?

ltratt commented on Objects/object.c:
```diff
@@ -1197,6 +1209,18 @@ _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
         return 0;
     }
 
+    if (Py_Py2xWarningFlag && PyUnicode_Check(name) 
```
Putting this check on every object is probably going to slow things down a lot. Should this go on the dict class/struct?

Eddy114514 replied:
@ltratt Thank you for pointing out, I think maybe I can adjust my approach by actually adding the attribute .sort to dict view object that do nothing and warn. 

Now moved to dictobject.c

nanjekyejoannah commented on Objects/dictobject.c:
```diff
@@ -5081,11 +5081,36 @@ static PyObject* dictkeys_reversed(_PyDictViewObject *dv, PyObject *Py_UNUSED(ig
 PyDoc_STRVAR(reversed_keys_doc,
 "Return a reverse iterator over the dict keys.");
 
+/*No-Operation for dictview.sort(), Just warn under -2 flag*/
```
Why do you keep defining new functions, motivation, I thought you are supposed to just warn in the existing functions? Maybe warn in the keys().... etc we don't implement functions just for warning. sort is applied on things we can iterate like lists, because dict.keys(),values(), items() now return objects, that is why it fails, so warnings just make sense outside sort.

We might have talked of alternatively checking at sort() and doing implied conversion before warning, but I think it makes more sense to warn in the dict operations. I think our documentation mentioning to explicitly warning on the sort method @ltratt  might have been an oversight, thoughts?

Eddy114514 replied:
@nanjekyejoannah I think just warn in `dict.key()` won’t fully capture the user’s intent (hard to distinguish whether the user just wants to iterate the view or actually intends to call .sort())
For example, suppose user write `for x in dict.key()`, it will get a false negative warning. 

And to warn in `.sort()` is also hard, since dict view object don't have such method attribute, it will raise attribute error before warning in `.sort()`. For the same reason, implied conversion is also hard because attribute error will raise first.

nanjekyejoannah replied:
I think you go inside sort() at the start do a check, convert, before the code continues, no?

I dont know if its wise to create dummy functions like this either, I would be happier defaulting to vague dict.x() warning. @ltratt ???

nanjekyejoannah replied:
We can have a quick zoom or if you are on campus, I will be available 9-12, then 3-5 tomorrow.

Eddy114514 replied:
@nanjekyejoannah I thought for `dict.keys().sort()`, any check or conversion in `sort()` won't process because attribute error will raise first unless we convert `dict.keys()` to list before `.sort` happen.  I am for available both time for zoom on tomorrow. Looking forward to discuss with you.  Thank you for reviewing.

Or I can try to modify tp_getattro (which is PyObject_GenericGetAttr) in dictobject.c, but it still need to return a callable object if we want to avoid run time error. If it is ok to raise an error, then it can just be a wrapper for PyObject_GenericGetAttr that warn.

ltratt commented:
@Eddy114514 How does this one relate to https://github.com/softdevteam/pygrate3/pull/27 ?

Eddy114514 commented:
> @Eddy114514 How does this one relate to #27 ?

@ltratt I think #27 is a general version to this PR. And my implementation here prevent the runtime attribute error of (.sort) and warn the specific action. So technically if both merged, `dict.keys().sort()` will raise 2 warning, 1 from .keys(), 1 from .sort().

ltratt commented:
Does that mean that we just want #27, or is #27 too broad or ... ?

Eddy114514 commented:
> Does that mean that we just want #27, or is #27 too broad or ... ?

@ltratt I think the problem of #27 is that it can't avoid run time failure, and warning can't even be issued in case `dict.key.sort` because program fail (due to attribute error) before warning in keys() actually raise up.

ltratt commented:
@Eddy114514 I think you're saying that #27 won't work? If so, should you close it?

IIUC this PR (#32) correctly it turns `dict.keys().sort()` into a no-op (with a warning) rather than an error?

Eddy114514 commented:
> @Eddy114514 I think you're saying that #27 won't work? If so, should you close it?
> 
> IIUC this PR (#32) correctly it turns `dict.keys().sort()` into a no-op (with a warning) rather than an error?

@ltratt Yeah, I think you are right that we should close #27 because pirate 2 can issue warning when calling keys, items, values

ltratt commented:
@Eddy114514 Can I check on this bit:

> IIUC this PR (https://github.com/softdevteam/pygrate3/pull/32) correctly it turns dict.keys().sort() into a no-op (with a warning) rather than an error?

Eddy114514 commented:
> @Eddy114514 Can I check on this bit:
> 
> > IIUC this PR (#32) correctly it turns dict.keys().sort() into a no-op (with a warning) rather than an error?

@ltratt I think you should be abled to check it. My implementation is available here https://github.com/Eddy114514/pygrate3/tree/dictsort. Thank you for reviewing.

Eddy114514 commented:
@ltratt @nanjekyejoannah After having consensus, I implemented the warning for .reverse() 

The situation is a little different here, as there is already a built in function for dict view object that is called __reversed__ which actually can reverse the internal values. So I just add an attribute for reverse and warn within.

ltratt commented:
@nanjekyejoannah Do you have any comments on this one?

nanjekyejoannah commented:
Am abit confused about what you mean by attribute here? If a reverse function already exists.

I agree on warning but now what unique difference do we have now? e.g do you want to warn that its reversing a view vs a list?

Another general point, it depends on the py3.x version, i realized some things were kinda really "rolled back"
in some versions. The point is we have to be careful or specific in the warning. I privately reached out for the full context

Eddy114514 commented:
> Am abit confused about what you mean by attribute here? If a reverse function already exists.
> 
> I agree on warning but now what unique difference do we have now? e.g do you want to warn that its reversing a view vs a list?
> 
> Another general point, it depends on the py3.x version, i realized some things were kinda really "rolled back" in some versions. The point is we have to be careful or specific in the warning. I privately reached out for the full context

 there is a function called `__reversed__`, so I just created an attribute like for .sort that point to `__reversed__`. So it avoid attribute error caused by calling `dict.key().reverse()`. 

Thanks for pointing out about versions difference in 3.x, I will focus on it to ensure the generality.

nanjekyejoannah commented:
OK, yes thats looks plausible, but the warning needs to hint on the supported reversed as you say and any other current 2.x vs 3.x context of the original reverse

Eddy114514 commented:
@nanjekyejoannah I have included warning that notice the support of `__reversed__` of dict view objects.  Also, after checking it seems that `__reversed__` is first introduced to py 3.8 (https://github.com/python/cpython/issues/77643). Should I avoid point to `__reversed__` to ensure version generality?

ltratt commented:
I think this one is ready now. @nanjekyejoannah  ?

nanjekyejoannah commented:
LGTM, lets merge this PR as is.

The warning is accurate for our current version of pygrate3 because this CPython 3.12 fork/mirror that we are building on but we have to create some disclaimer because this feature in 3.8 and prior reversed() isn't supported so the warning would be different. We can handle this in a future iteration if we ever decide to rethink addressing py3.x versions which don't have the rollback

Eddy114514 committed:
Add warning for sort in dict.keys()/values()/items()
Add warning for reverse
Add attribute sort and reverse to avoid run time
attribute error.
